File: check_valid_utf8.pl
#!/usr/bin/env perl
#
# Takes a list of files on the command line and checks for valid
# UTF-8 data. Used for checking .po files.
#
# Copyright (C) 2009 Tobias Quathamer <toddy@debian.org>
# Released under the GPL version 2 or later.

use strict;
use warnings;

my $exit_status = 0;

foreach my $filename (@ARGV) {
  my $content_type_checked = 0;
  open FILE, "< $filename";
  while (<FILE>) {
    # Check for valid UTF-8 encoding
    unless (m/\A(
      [\x09\x0A\x0D\x20-\x7E]            # ASCII
    | [\xC2-\xDF][\x80-\xBF]             # non-overlong 2-byte
    |  \xE0[\xA0-\xBF][\x80-\xBF]        # excluding overlongs
    | [\xE1-\xEC\xEE\xEF][\x80-\xBF]{2}  # straight 3-byte
    |  \xED[\x80-\x9F][\x80-\xBF]        # excluding surrogates
    |  \xF0[\x90-\xBF][\x80-\xBF]{2}     # planes 1-3
    | [\xF1-\xF3][\x80-\xBF]{3}          # planes 4-15
    |  \xF4[\x80-\x8F][\x80-\xBF]{2}     # plane 16
    )*\z/x) {
      # Found invalid characters for UTF-8
      printf("Error in file %s at line number %d:\n", $filename, $.);
      # Show the line with the error
      print;
      $exit_status = 1;
      # Skip the rest of the current file
      last;
    }
    # Check that the Content-Type header field is set correctly.
    if (!$content_type_checked && /Content-Type: text\/plain; charset=UTF-8/) {
      $content_type_checked = 1;
    }
  }
  unless ($content_type_checked) {
    printf("Error in file %s:\n", $filename);
    print("Could not detect correct Content-Type header field.\n");
    $exit_status = 1;
  }
  close FILE;
}

exit($exit_status);

File: scripts/check_valid_utf8.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: 2016 Dr. Tobias Quathamer <toddy@debian.org>
#
# SPDX-License-Identifier: LGPL-2.1-or-later

"""
Takes a directory on the command line and checks for valid
UTF-8 data in the .po files therein.
"""

import pathlib
import re
import sys

# Get the directory to check
if len(sys.argv) != 2:
    print("Error: Provide the directory to check.", file=sys.stderr)
    sys.exit(1)
directory = sys.argv[1]

# Assume that every file is valid
exit_status = 0

po_files = [f for f in pathlib.Path().glob(directory + "*po")]

# Cycle through all .po files to check for valid UTF-8 encoding
for filename in po_files:
    # Open the file for reading in binary mode
    with open(filename, "rb") as pofile:
        # The "Content-Type" header has not been seen yet
        charset_utf8_seen = False
        # Read all lines to check for Content-Type header
        for line in pofile:
            # Try to decode binary data to UTF-8
            try:
                utf8 = line.decode(encoding="utf-8", errors="strict")
            except UnicodeError as error:
                print(
                    f"UTF-8 encoding error in file {filename}: {error.reason} (position {error.start})",
                    file=sys.stderr,
                )
                print(f"Binary data: {line}", file=sys.stderr)
                exit_status = 1
                break
            if re.search(r"Content-Type: text/plain; charset=UTF-8", utf8):
                charset_utf8_seen = True
        # The whole file has been read, the content type should have
        # been detected now. Otherwise, it's an error.
        if not charset_utf8_seen:
            print(
                f"Error in file {filename}: could not detect UTF-8 Content-Type header",
                file=sys.stderr,
            )
            exit_status = 1
            break

sys.exit(exit_status)
